### ISSUE TITLE
Switching from MIT to GNU AGPLv3

### ISSUE THREAD

**Issue by kerkmann** [fwcd/kotlin-language-server#211]
I think that switching to the Affero GPL is a giant step forward.
Open source is very import and software piracy is quite commonly with the MIT.
So the GPL and AGPL protect the open source system and have some powerful copyleft.
AGPL is quite popular on project like owncloud/Nextcloud, proxmox, estudio, mastodon, mongodb, Snap!.
I think it's very important to use the GNU Affero General Public License instead of the MIT. Just to be sure that this code will be open source forever and not behind some legacy code. :)

And yes it's possible to switch from MIT to [AGPLv3](https://www.gnu.org/licenses/agpl-3.0.en.html). :)

Some useful sources:
https://tldrlegal.com/license/gnu-affero-general-public-license-v3-(agpl-3.0)
https://exygy.com/blog/which-license-should-i-use-mit-vs-apache-vs-gpl/

**Comment by fwcd** [comment]
I have been thinking about switching to a GPL-style license for a while already, so I do support this idea! Not quite sure whether we should go with AGPLv3 or GPLv3 though. Being able to embed the language server into GPL-contexts might be nice.

**Comment by trygveaa** [comment]
Wouldn't this require permission from all the contributors?

**Comment by fwcd** [comment]
@trygveaa MIT, the license under which all contibutions so far have been licensed, is very permissive and can be migrated to a less permissive license like GPL/AGPL, while still complying with MIT's terms.

**Comment by kerkmann** [comment]
AGPL is quite similar to GPL.
"It just adds some network layers". :D
It's a new form of GPL that was designed for applications that run through a network.
It's possible to use a GPL client with the AGPL server. The client just need to clarify that the backend code is using this source.
Like mastodon did that long time ago.
https://niu.moe/about
On the footer you see a "see source code" buttony that redirects to the GitHub page.
That alle what a client needs to do, and that's the difference between GPL and AGPL.
You could run an GPL backend and nobody will know that. :)
That's why AGPL is what I think the better solution. :)

If you write an client, you don't need to apply the AGPL license, you can also use GPL.
It's just important to make the AGPL code available. ;)

**Comment by fwcd** [comment]
Note that there may be more things to consider before/if we switch to (A)GPL. This is of course a philosophical debate, but we might risk losing contributors [upstreaming their work from closed systems](https://github.com/fwcd/kotlin-language-server/issues/103#issuecomment-494350487) incorporating this language server.

cc @vasiltabakov

**Comment by kerkmann** [comment]
Which parts are infected?
Or is it only this PR?

**Comment by fwcd** [comment]
The reason I'm being hesitant is not due to certain parts being affected (everything currently in the repository is MIT-licensed). There are [advantages to permissive licensing](https://www.reddit.com/r/linux/comments/g55t4t/why_the_mit_license_is_much_more_used_then_gpl3/fo1v6xb/) too and I'd like to hear more opinions from the contributors before we move on here.

**Comment by fwcd** [comment]
More things to consider:

- Do we really want **license fragmentation**? GPL'ing the repository would not nullify the MIT license on anything currently in the repository. Additionally, it might cause people unhappy with a GPL-style license to fork the repository into an MIT version.
- GPL code **cannot** be used in projects licensed e.g. under EPL ([JDT language server](https://github.com/eclipse/eclipse.jdt.ls/)) oder Apache 2.0 ([NetBeans](https://github.com/apache/netbeans)). This would prevent the hypothetical scenario that we may want to upstream code from the language server to these projects.

**Comment by kerkmann** [comment]
Yes I understand what you think. But especially for this project some arguments are not very weighted.
Arguments like the warranty, or arguments like using libraries in the games industries doesn't matter for such an project. (My personal opinion). 
We should keep the scope in mind and decide for which scenario is for.
As I can say, it's not a fancy library that will be used in some medical/military/game projects. :)
But you're right, I'm completely with you and we should hear the meanings of each contributor. :)

I thought the same about EPL, but I think we should think about where we want to go.
What plans do we have? Is it possible to work together with JDT language server, or should be be an different branch.

For me I'm very tied up with my opinion and I would prefer a AGPL license more than MIT.
Just because this license can work perfectly in many situations and I love the idea behind that.
And I think that the MIT is too harmful, there are better licenses. ^^
But as you sad, It's important to ask any contributor about their opinion and think about all pros, cons and where we want to go.

Do you have a good idea to talk with the other contributor or should this issue just be overwhelmed?

**Comment by DonnieWest** [comment]
As much as I like the broad category of GPL licenses, I don't think this is a good idea for this project.

1. We're bikeshedding on philosophical differences
2. The current choice of license has worked for us thus far without negatively effecting the project
3. Changing the license is unnecessary work
  a. We need determine it's going to work for the project without alienating current and future contributors
  b. We need to determine what form of license among many to choose when the current one works just fine

The decision of license was made on the outset of the project and shouldn't be changed without broad consensus and a compelling case beyond philosophical musings. Conversation about "the scope" of the project was essentially already determined.

Unless @fwcd wants to exercise BDFL role and switch the license unilaterally, which I frankly wouldn't be opposed to

tl;dr - more work for minimal benefit

**Comment by kerkmann** [comment]
@fwcd @DonnieWest I think both of you are totally right. :)
The decision was already made and there is no step forward to find a solution right now, for something that is working already. :)
I just thought it would be a great idea, but as I saw the opinions are different, it was quite sure and I don't wanna upset someone. ^^"

I'll close that issue, thanks for your feedback. :)
